src/main/java/com/puppycrawl/tools/checkstyle/checks/coding/ConstructorsDeclarationGroupingCheck.java

```diff
@@ -0,0 +1,104 @@
+///////////////////////////////////////////////////////////////////////////////////////////////
+// checkstyle: Checks Java source code and other text files for adherence to a set of rules.
+// Copyright (C) 2001-2024 the original author or authors.
+//
+// This library is free software; you can redistribute it and/or
+// modify it under the terms of the GNU Lesser General Public
+// License as published by the Free Software Foundation; either
+// version 2.1 of the License, or (at your option) any later version.
+//
+// This library is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+// Lesser General Public License for more details.
+//
+// You should have received a copy of the GNU Lesser General Public
+// License along with this library; if not, write to the Free Software
+// Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA  02111-1307  USA
+///////////////////////////////////////////////////////////////////////////////////////////////
+
+package com.puppycrawl.tools.checkstyle.checks.coding;
+
+import java.util.HashMap;
+import java.util.Map;
+
+import com.puppycrawl.tools.checkstyle.StatelessCheck;
+import com.puppycrawl.tools.checkstyle.api.AbstractCheck;
+import com.puppycrawl.tools.checkstyle.api.DetailAST;
+import com.puppycrawl.tools.checkstyle.api.TokenTypes;
+
+/**
+ * <p>
+ * Checks that all constructors are grouped together.
+ * If there is anything between overloaded constructors ( expect comments )
+ * then this check will give an error.
+ * </p>
+ * <p>
+ * Parent is {@code com.puppycrawl.tools.checkstyle.TreeWalker}
+ * </p>
+ * <p>
+ * Violation Message Keys:
+ * </p>
+ * <ul>
+ * <li>
+ * {@code constructors.declaration.grouping}
+ * </li>
+ * </ul>
+ *
+ * @since 10.16.0
+ */
+
+@StatelessCheck
+public class ConstructorsDeclarationGroupingCheck extends AbstractCheck {
+
+    /**
+     * A key is pointing to the warning message text in "messages.properties"
+     * file.
+     */
+    public static final String MSG_KEY = "constructors.declaration.grouping";
+
+    /**
+     * Specifies different Object Blocks scope.
+     */
+    private final Map<DetailAST, Integer> allObjBlocks = new HashMap<>();
```

Comment:
This is a state.
If we can not avoid state ( fields of class) , we should clean/reset fields at the begging of each file validation. Same instance of Check is reused over all files to validate.

 If you see no way to make it stateless, it is ok. You can just explain a reason on why stateless is not possible, and stay with stateful.

Reply:
> This is a state. If we can not avoid state ( fields of class) , we should clean/reset fields at the begging of each file validation. Same instance of Check is reused over all files to validate.

Done. Used the `beginTree()` method to achieve this:

https://github.com/checkstyle/checkstyle/blob/cadb0dc3a4635dd53eaa24ad6d41fd6e2cef7fd2/src/main/java/com/puppycrawl/tools/checkstyle/checks/coding/ConstructorsDeclarationGroupingCheck.java#L83-L86


---

> If you see no way to make it stateless, it is ok. You can just explain a reason on why stateless is not possible, and stay with stateful.

I guess we can't make this check stateless because the check requires tracking the scope of the overloaded constructors, so we need a global variable to keep track of this. A local variable will not work in this situation. Here is the example of what I'm trying to say:

https://github.com/checkstyle/checkstyle/blob/cadb0dc3a4635dd53eaa24ad6d41fd6e2cef7fd2/src/test/resources/com/puppycrawl/tools/checkstyle/checks/coding/constructorsdeclarationgrouping/InputConstructorsDeclarationGrouping.java#L29-L53

@romani

Reply:
Ok